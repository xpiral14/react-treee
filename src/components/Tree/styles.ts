import styled from "styled-components";

export const Container = styled.ul`
  &,
  & ul {
    list-style: none;
    margin: 0;
    padding: 0;
  }
  & ul {
    padding-left: 20px;
  }
  & li {
    line-height: 30px;
    color: #369;
    font-weight: bold;
    border-left: 2px solid rgb(100, 100, 100);
    margin-left: 20px;

    > div {
      margin-left: 38px;
      display: flex;
      align-items: flex-start;
    }
    /* padding: 0 20px; */
    :before {
      position: relative;
      top: 1.5em;
      padding: 0 10px;
      height: 30px;
      width: 10px;
      margin-left: 7px;
      border-bottom: 2px solid rgb(100, 100, 100);
      content: "";
      display: inline-block;
      left: -7px;
    }
  }
`;
