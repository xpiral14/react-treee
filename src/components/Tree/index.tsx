import TreeNode from "../TreeNode";
import { Container } from "./styles";
export interface Node {
  id: string;
  name: string;
  children: Record<string, Node>;
  level: number;
}

export interface NodeWithParent extends Node {
  parent: {
    id: string;
    name: string;
  };
}
export interface ITreeProps {
  nodes: Record<string, Node>;
  selecteds?: string[];
  onSelect?: (ids: string[]) => void;
  onUnselect?: (ids: string[]) => void;
}


export default function Tree({
  nodes,
  selecteds,
  onSelect,
  onUnselect,
}: ITreeProps) {
  return (
    <Container>
      {Object.values(nodes).map((node) => (
        <TreeNode
          key={node.id}
          selecteds={selecteds}
          node={node}
          onSelect={onSelect}
          onUnselect={onUnselect}
          parentNodes={{}}
        />
      ))}
    </Container>
  );
}
