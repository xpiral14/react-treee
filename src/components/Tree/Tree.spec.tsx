/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */
import React from "react";
import { cleanup, render, screen } from "@testing-library/react";

import { create } from "react-test-renderer";
import Tree from "./index";

import { configure, mount, shallow } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";

configure({ adapter: new Adapter() });

const fakeTree = {
  "0": {
    id: "1",
    name: "Fátima Araújo",
    level: 0,
    children: {
      "0": {
        id: "1-1",
        name: "Samuel Reis",
        level: 1,
        children: {},
      },
      "1": {
        id: "1-2",
        name: "Sabrina Reis",
        level: 1,
        children: {},
      },
      "3": {
        id: "1-3",
        name: "Alexandre Lins",
        level: 1,
        children: {
          "0": {
            id: "1-3-1",
            name: "Emilly sofia",
            level: 2,
            children: {},
          },
        },
      },
    },
  },
};

afterAll(cleanup);
describe("<Tree />", () => {
  test("snapshot", () => {
    const component = create(<Tree nodes={fakeTree} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("if can uncollapse chidlren nodes", () => {
    const onSelect = jest.fn();
    const onUnselect = jest.fn();
    const { container } = render(
      <Tree nodes={fakeTree as any} onSelect={onSelect} onUnselect={onUnselect} />
    );

    (container.querySelector("#button_1-3") as any)?.click();

    screen.queryByText("Emilly sofia")?.click();

    expect(onSelect).toBeCalledTimes(1);
  });

  test("if has correct number of childs", () => {
    const wrapper = mount(<Tree nodes={fakeTree} />);

    expect(wrapper.find("li").length).toEqual(4);
  });

  test("if can unselect nodes at tree", () => {
    const onSelect = jest.fn((data) => {console.log(data)});
    const onUnselect = jest.fn();

    const wrapper = mount(
      <Tree
        nodes={fakeTree as any}
        onSelect={onSelect}
        onUnselect={onUnselect}
        selecteds={["1", "1-1", "1-2", "1-3", "1-3-1"]}
      />
    );
    wrapper.find("#checkbox_1")?.simulate("click");
    expect(onSelect).toBeCalledTimes(2);

    expect(onSelect).toBeCalledWith(expect.arrayContaining(["1", "1-1", "1-2", "1-3", "1-3-1"]));
  });
});
