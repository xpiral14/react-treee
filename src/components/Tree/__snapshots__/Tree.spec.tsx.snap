// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tree /> snapshot 1`] = `
<ul
  className="sc-eCImPb jKSZlF"
>
  <li
    className="sc-hKwDye"
  >
    <div>
      <button
        id="button_1"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height={20}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
          />
        </svg>
      </button>
      <div
        className="sc-bdvvtL DfnJp"
        id="1"
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
      >
        <label
          htmlFor="checkbox_1"
        >
          <input
            id="checkbox_1"
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="sc-gsDKAQ eLUAnk"
          >
            <div
              className="sc-dkPtRN ktxWBh"
            />
          </div>
          Fátima Araújo
        </label>
      </div>
    </div>
    <ul>
      <li
        className="sc-hKwDye"
      >
        <div>
          <div
            className="sc-bdvvtL DfnJp"
            id="1-1"
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
          >
            <label
              htmlFor="checkbox_1-1"
            >
              <input
                id="checkbox_1-1"
                onChange={[Function]}
                type="checkbox"
              />
              <div
                className="sc-gsDKAQ eLUAnk"
              >
                <div
                  className="sc-dkPtRN ktxWBh"
                />
              </div>
              Samuel Reis
            </label>
          </div>
        </div>
      </li>
    </ul>
    <ul>
      <li
        className="sc-hKwDye"
      >
        <div>
          <div
            className="sc-bdvvtL DfnJp"
            id="1-2"
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
          >
            <label
              htmlFor="checkbox_1-2"
            >
              <input
                id="checkbox_1-2"
                onChange={[Function]}
                type="checkbox"
              />
              <div
                className="sc-gsDKAQ eLUAnk"
              >
                <div
                  className="sc-dkPtRN ktxWBh"
                />
              </div>
              Sabrina Reis
            </label>
          </div>
        </div>
      </li>
    </ul>
    <ul>
      <li
        className="sc-hKwDye"
      >
        <div>
          <button
            id="button_1-3"
            onClick={[Function]}
          >
            <svg
              fill="currentColor"
              height={20}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
              />
            </svg>
          </button>
          <div
            className="sc-bdvvtL DfnJp"
            id="1-3"
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
          >
            <label
              htmlFor="checkbox_1-3"
            >
              <input
                id="checkbox_1-3"
                onChange={[Function]}
                type="checkbox"
              />
              <div
                className="sc-gsDKAQ eLUAnk"
              >
                <div
                  className="sc-dkPtRN ktxWBh"
                />
              </div>
              Alexandre Lins
            </label>
          </div>
        </div>
      </li>
    </ul>
  </li>
</ul>
`;
